tapchannel/aux_sweeper.go

```diff
@@ -360,6 +384,190 @@ func commitRevokeSweepDesc(keyRing *lnwallet.CommitmentKeyRing,
+	// output to the packet to complete it.
+	assetProof := assetOutput.Proof.Val
+	inclusionProof := assetProof.InclusionProof
+	vOut := &tappsbt.VOutput{
+		Amount:       assetProof.Asset.Amount,
+		AssetVersion: assetProof.Asset.Version,
+		Type: func() tappsbt.VOutputType {
+			if assetProof.SplitRootProof == nil {
+				return tappsbt.TypeSimple
+			}
+			return tappsbt.TypeSplitRoot
+		}(),
+		Interactive:             true,
+		AnchorOutputIndex:       inclusionProof.OutputIndex,
+		AnchorOutputInternalKey: inclusionProof.InternalKey,
+		//nolint:ll
+		AnchorOutputTapscriptSibling: inclusionProof.CommitmentProof.TapSiblingPreimage,
+		ScriptKey:                    assetProof.Asset.ScriptKey,
+		ProofSuffix:                  &assetProof,
+	}
+
+	// While we're here, we'll also replace the transaction stored in the
+	// proof with the correct one.
+	vOut.ProofSuffix.AnchorTx = *commitTx
+
+	// Finally, we'll make the proper control block to spend the OP_TRUE
+	// funding output.
+	fundingWitness, err := fundingSpendWitness().Unpack()
+	if err != nil {
+		return fmt.Errorf("unable to make funding witness: %w", err)
+	}
+
+	// There is always only a single input, which is the funding output.
+	const inputIndex = 0
+	vOut.Asset.UpdateTxWitness(0, fundingWitness)
+
+	// Finally, we'll set the delivery address to the default courier, so
+	// we publish the proof in the specified Universe.
+	vOut.ProofDeliveryAddress = courierAddr
+
+	vPkt.Outputs = append(vPkt.Outputs, vOut)
+	vPktsByAssetID[assetID] = vPkt
+
+	return nil
+}
+
+// importCommitTx imports the commitment transaction into the wallet. This is
+// called after a force close to ensure that we can properly spend outputs
+// created by the commitment transaction at a later step.
+func (a *AuxSweeper) importCommitTx(req lnwallet.ResolutionReq,
+	commitState *cmsg.Commitment, fundingInfo *cmsg.OpenChannel) error {
+
+	// To start, we'll re-create vPackets for all of the outputs of the
+	// commitment transaction.
+	//
+	// We'll use the fundingInfo proofs to create a vIn for each of them.
+	fundingInputProofs := make(map[asset.ID]*proof.Proof)
+	for _, fundingInput := range fundingInfo.FundedAssets.Val.Outputs {
+		inputProof := &fundingInput.Proof.Val
+		fundingInputProofs[inputProof.Asset.ID()] = inputProof
+	}
+
+	// With the funding proof for each asset ID known, we can now make the
+	// vPackets for each of the outputs.
+	vPktsByAssetID := make(map[asset.ID]*tappsbt.VPacket)
+	for _, localAsset := range commitState.LocalAssets.Val.Outputs {
+		err := assetOutputToVPacket(
+			fundingInputProofs, vPktsByAssetID, localAsset,
+			&a.cfg.ChainParams, req.CommitTx,
+			a.cfg.DefaultCourierAddr,
+		)
+		if err != nil {
+			return err
+		}
+	}
+	for _, remoteAsset := range commitState.RemoteAssets.Val.Outputs {
+		err := assetOutputToVPacket(
+			fundingInputProofs, vPktsByAssetID, remoteAsset,
+			&a.cfg.ChainParams, req.CommitTx,
+			a.cfg.DefaultCourierAddr,
```

Comment:
In the co-op close I added the desired delivery address of each party to the shutdown message. Since there is no interactivity here, I think we should perhaps negotiate that upfront, during channel funding?
Otherwise if I use a non-standard universe and force close, you as my peer wouldn't know what universe to fetch the proof from.

Reply:
Yeah makes sense. 

FWIW, I added this here mainly so I could add an assertion that it was properly uploaded to the Universe. I saw in another commit, you started to do validation in the porter before moving on. Without that, I would have a transfer, but then universe upload fails as an inclusion proof was wrong or something. 

In the latest diff, both sides are able to complete+import the close proof themselves. Then the responder also complete the proof for the funding transaction itself. This would move to later live in the `chanvalidate` package. so as soon as the funding txn is confirmed, we complete the proof then try to validate before using it at all.